<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Work Package 3: Heterogeneity – DP-Next</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-8fd1faf6fd93e4e6d207ae0f9db51d78.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed fullcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">DP-Next</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Overview</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-work-packages" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Work packages</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-work-packages">    
        <li>
    <a class="dropdown-item" href="./wp1/index.html">
 <span class="dropdown-text">WP1: Better research in less time</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./wp2.html">
 <span class="dropdown-text">WP2: Risk Prediction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./wp3.html">
 <span class="dropdown-text">WP3: Heterogeneity</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./wp4.html">
 <span class="dropdown-text">WP4: Intervention Development</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./who-we-are.html"> 
<span class="menu-text">Who We Are</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./publications.html"> 
<span class="menu-text">Publications</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./support.html"> 
<span class="menu-text">Support</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://github.com/dp-next/" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Work Package 3: Heterogeneity</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>WP3 will characterise heterogeneity among people in early middle age with HbA1c defined pre-diabetes. The aim is to identify a set of easily obtainable biomarkers that can optimally distinguish those with high probability of stable pre-diabetes/remission from those at highest risk of progression to diabetes.</p>
<section id="wp-3.1-risk-clustering-and-long-term-prediction" class="level2">
<h2 class="anchored" data-anchor-id="wp-3.1-risk-clustering-and-long-term-prediction">WP 3.1 Risk clustering and long-term prediction</h2>
<p>WP3.1 will investigate to which degree detailed biological data (e.g.&nbsp;genetics, omics, health behavioural data) can improve risk prediction within a Danish (<a href="https://bmcpublichealth.biomedcentral.com/articles/10.1186/1471-2458-12-1078">ADDITION-PRO</a>) and a Greenlandic (<a href="https://www.tandfonline.com/doi/full/10.1080/22423982.2019.1709257">Greenland Health Surveys</a>) context. Both cohorts recruited participants more than a decade ago and examined a wide set of cardiometabolic and genetic risk factors, health behaviours and biomarkers, including fat distribution measures and physical activity measures from week long combined heart rate/accelerometer readings. We will link these cohorts to the register-based risk predictions from WP2 (backdated to each cohort baseline), and examine the added predictive value of individual biological risk indicators and of risk factor clusters. The availability of more than a decade of accrued follow-up for diabetes incidence in these cohorts will enable us to obtain a global indication of diabetes risk clusters within the first year of DP-Next.</p>
</section>
<section id="wp-3.2-deep-phenotyping-of-hba1c-defined-pre-diabetes" class="level2">
<h2 class="anchored" data-anchor-id="wp-3.2-deep-phenotyping-of-hba1c-defined-pre-diabetes">WP 3.2 Deep Phenotyping of HbA1c defined pre-diabetes</h2>
<section id="hypothesis" class="level3">
<h3 class="anchored" data-anchor-id="hypothesis">Hypothesis:</h3>
<p>Among individuals with routinely identified prediabetes based on HbA1c, deep phenotyping with non-invasive methods will identify subgroups at the highest risk of progression to T2D and subgroups likely to maintain prediabetes or achieve remission; beyond the predictive ability of age, sex, HbA1c and BMI.</p>
</section>
<section id="background" class="level3">
<h3 class="anchored" data-anchor-id="background">Background:</h3>
<p>Although elevated blood glucose precedes T2D onset by over a decade<span class="citation" data-cites="Tabak2009">(<a href="#ref-Tabak2009" role="doc-biblioref">1</a>)</span>, many people with slightly elevated HbA1c do not progress to diabetes<span class="citation" data-cites="Vistisen2019">(<a href="#ref-Vistisen2019" role="doc-biblioref">2</a>)</span>. Depending on the criteria for prediabetes, 10-40% develop T2D within 1-5 years<span class="citation" data-cites="Hong2016 Larsson2000 Tabak2012">(<a href="#ref-Hong2016" role="doc-biblioref">3</a>–<a href="#ref-Tabak2012" role="doc-biblioref">5</a>)</span> and 17-45% revert to normoglycaemia<span class="citation" data-cites="Vistisen2019">(<a href="#ref-Vistisen2019" role="doc-biblioref">2</a>)</span>, pointing to significant heterogeneity in diabetes risk among individuals with prediabetes<span class="citation" data-cites="Hulman2017 Hulman2018 Hulman2018a">(<a href="#ref-Hulman2017" role="doc-biblioref">6</a>–<a href="#ref-Hulman2018a" role="doc-biblioref">8</a>)</span>. A relatively recent cluster-driven diabetes subclassification delineated two subgroups with relatively young age of diabetes onset and absence of marked obesity and low insulin secretion<span class="citation" data-cites="Ahlqvist2018">(<a href="#ref-Ahlqvist2018" role="doc-biblioref">9</a>)</span>, which map to genetic, metabolic and phenotypic characteristics already present in pre-diabetes. Notably, a recent study identified elevated liver fat and diminished beta-cell function as the most predictive factors for progressing to T2D in people with pre-diabetes<span class="citation" data-cites="Wagner2021">(<a href="#ref-Wagner2021" role="doc-biblioref">10</a>)</span>, highlighting the need to extend precision approaches for diabetes prevention beyond genetic predisposition and insulin resistance. The current vision for precision prevention integrates data from genetic, biological (multi-omic), health behavioural and psycho-social components; however, few cohorts have these data in a contemporary population defined in a clinical context.</p>
</section>
<section id="strategic-vision" class="level3">
<h3 class="anchored" data-anchor-id="strategic-vision">Strategic Vision:</h3>
<p>WP3 establishes such a cohort across Steno Centers with long-term academic and strategic visions. The strategic vision is to establish the facilities, expertise and capacity across all Steno Centers to carry out multi-center clinical studies with a standardised protocol and management/data infrastructure. We also want to prepare the centers for the standardised application of deep phenotyping close to clinical practice by applying a high degree of standardisation, openness and reproducibility (<a href="https://steno-aarhus.github.io/dp-next/wp1/">WP1</a>).</p>
</section>
<section id="academic-objectives" class="level3">
<h3 class="anchored" data-anchor-id="academic-objectives">Academic Objectives:</h3>
<p>Academically, we describe a core protocol, which will be carried out within the DP-Next budget and timeframe, and an extended protocol, which describes our ambitions for deeper phenotyping and longer follow-up. Funding for the extended protocol will be sought separately on a project basis and will build on the capacity established in the core protocol. Connection and coordination will be sought with new Danish phenotyping initiatives currently underway or in the planning process (e.g.&nbsp;PRECISE).</p>
</section>
<section id="who-will-be-included" class="level3">
<h3 class="anchored" data-anchor-id="who-will-be-included">Who will be included:</h3>
<p>In the core protocol we will identify a cohort of 1000 individuals aged 40 to 55 years who have recently (within the past 12 months) had a routine HbA1c measurement in the pre-diabetic range (42-47 mmol/mol) and are not using any glucose lowering medication. This age range was chosen to maximise life-long impact on diabetes risk. Sample size calculations use our recent analysis of HbA1c defined (pre)diabetes in Denmark<span class="citation" data-cites="Nicolaisen2023">(<a href="#ref-Nicolaisen2023" role="doc-biblioref">11</a>)</span>, showing that 25 to 40% of Danish residents aged 40-55 had at least 1 HbA1c measurement in 2018. In our target age range pre-diabetes incidence was 9/1000 person years, and the subsequent incidence of diabetes was 7/1000, and hence an expected cumulative T2D incidence of 22% in a median of 3.5 years of follow-up. Logistic regression based power calculations show that a cohort of 1050 individuals gives us 0.9 power at a 0.05 significance level to detect an odds ratio of at least 1.28 per standard deviation in a continuous exposure variable.</p>
</section>
<section id="recruitment" class="level3">
<h3 class="anchored" data-anchor-id="recruitment">Recruitment:</h3>
<p>Participants will be recruited in Odense (350), Aarhus (350), Aalborg (150), Greenland (50) and the Faroe Islands (50). We intend to include 100 individuals of Greenlandic ancestry, 50 living in Denmark and 50 living in Greenland. Participants will be recruited based on data from routine HbA1c checks, accessed through the LABKA registers. Consequently the clinical decisions prompting a HbA1c measurement will form part of the selection process, reflecting current clinical practice.</p>
</section>
<section id="core-protocol" class="level3">
<h3 class="anchored" data-anchor-id="core-protocol">Core Protocol:</h3>
<p>Participants will be invited to answer an online questionnaire (medical and family history, sociodemographic data, health literacy, food habits and cravings , personality traits such as willingness to take risks, quality of life, self-perceived mental stress, depression, anxiety, sleep apnea scores) followed by a 4-hour clinical examination at one of the Steno Diabetes Centers. The programme includes: anthropometric measures (height, weight, waist-hip ratio, and DEXA-scan for body composition), a 5-point oral glucose tolerance test to estimate insulin sensitivity and beta-cell function, liver elastography to estimate liver fat and liver stiffness, blood pressure, heart rate variability and pulse-wave velocity for arterial stiffness estimation. Participants will measure physical activity with a combined heart rate monitor / accelerometer during 7 days following the visit. Blood samples will be analysed for lipids and HbA1c and processed for biobanking of plasma, serum and DNA. The biobank will further include urine, saliva and faeces samples. All further measurements will be in biobank samples as part of the extended protocol.</p>
<p>The primary outcome for the core protocol is incident Type 2 Diabetes. The LABKA register will identify incident T2D (based on routine HbA1c) supplemented with a study HbA1c measurement at the end of the DP-Next project. The primary statistical analysis will use unsupervised Latent Class Analysis<span class="citation" data-cites="Weller2020">(<a href="#ref-Weller2020" role="doc-biblioref">12</a>)</span> to map the heterogeneity across all measured variables. LCA will be carried out at two levels: a full model including all available data and a minimal model using as few data as possible with limited loss of strength. Modelling studies show that our sample of 1000 is sufficient to identify patterns and clusters in most data sets.</p>
</section>
<section id="extended-protocol" class="level3">
<h3 class="anchored" data-anchor-id="extended-protocol">Extended Protocol:</h3>
<p>Our ambition is to extend the WP3 protocol beyond the budget and time of DP-Next. Additional funding will be sought to expand the depth of the cardiometabolic phenotyping under real life settings, including a 24-hour ambulatory blood pressure, cardiorespiratory monitoring, and continuous glucose monitoring, with simultaneous app-based diet diaries. We further have the ambition to invite a subset of ~300 participants for a second examination with optical coherence tomography to detect early retinal changes, a single breath-hold MR scan of the liver, and an OGTT during short-term oral steroid treatment. We intend to expand the project carrying out a comprehensive set of omics analyses in biobank materials, including genotyping, proteomics, metabolomics and metagenomic characterisation of faeces and saliva as well as a set of targeted biomarkers covering mechanisms such as endothelial function, liver function and auto-immunity (GAD and islet cell antibodies). Our ambition is to extend the follow-up beyond the DP-Next project period, extending the accrual not only of diabetes cases but also of secondary endpoints such as cardiovascular events.</p>



</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body" role="list">
<div id="ref-Tabak2009" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Tabák AG, Jokela M, Akbaraly TN, Brunner EJ, Kivimäki M, Witte DR. <a href="https://doi.org/10.1016/S0140-6736(09)60619-X">Trajectories of glycaemia, insulin sensitivity, and insulin secretion before diagnosis of type 2 diabetes: <span>An</span> analysis from the <span>Whitehall II</span> study</a>. Lancet (London, England). 2009 Jun;373(9682):2215–21. </div>
</div>
<div id="ref-Vistisen2019" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Vistisen D, Kivimäki M, Perreault L, Hulman A, Witte DR, Brunner EJ, et al. <a href="https://doi.org/10.1007/s00125-019-4895-0">Reversion from prediabetes to normoglycaemia and risk of cardiovascular disease and mortality: <span>The Whitehall II</span> cohort study</a>. Diabetologia. 2019 Aug;62(8):1385–90. </div>
</div>
<div id="ref-Hong2016" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Hong J-L, McNeill AM, He J, Chen Y, Brodovicz KG. <a href="https://doi.org/10.1002/pds.4007">Identification of impaired fasting glucose, healthcare utilization and progression to diabetes in the <span>UK</span> using the <span>Clinical Practice Research Datalink</span> (<span>CPRD</span>)</a>. Pharmacoepidemiology and Drug Safety. 2016 Dec;25(12):1375–86. </div>
</div>
<div id="ref-Larsson2000" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Larsson H, Lindgärde F, Berglund G, Ahrén B. <a href="https://doi.org/10.1007/s001250051516">Prediction of diabetes using <span>ADA</span> or <span>WHO</span> criteria in post-menopausal women: A 10-year follow-up study</a>. Diabetologia. 2000 Oct;43(10):1224–8. </div>
</div>
<div id="ref-Tabak2012" class="csl-entry" role="listitem">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Tabák AG, Herder C, Rathmann W, Brunner EJ, Kivimäki M. <a href="https://doi.org/10.1016/S0140-6736(12)60283-9">Prediabetes: A high-risk state for diabetes development</a>. Lancet (London, England). 2012 Jun;379(9833):2279–90. </div>
</div>
<div id="ref-Hulman2017" class="csl-entry" role="listitem">
<div class="csl-left-margin">6. </div><div class="csl-right-inline">Hulman A, Simmons RK, Vistisen D, Tabák AG, Dekker JM, Alssema M, et al. <a href="https://doi.org/10.1007/s12020-016-1126-z">Heterogeneity in glucose response curves during an oral glucose tolerance test and associated cardiometabolic risk</a>. Endocrine. 2017 Feb;55(2):427–34. </div>
</div>
<div id="ref-Hulman2018" class="csl-entry" role="listitem">
<div class="csl-left-margin">7. </div><div class="csl-right-inline">Hulman A, Vistisen D, Glümer C, Bergman M, Witte DR, Færch K. <a href="https://doi.org/10.1007/s00125-017-4468-z">Glucose patterns during an oral glucose tolerance test and associations with future diabetes, cardiovascular disease and all-cause mortality rate</a>. Diabetologia. 2018 Jan;61(1):101–7. </div>
</div>
<div id="ref-Hulman2018a" class="csl-entry" role="listitem">
<div class="csl-left-margin">8. </div><div class="csl-right-inline">Hulman A, Witte DR, Vistisen D, Balkau B, Dekker JM, Herder C, et al. <a href="https://doi.org/10.2337/dc18-0279">Pathophysiological <span>Characteristics Underlying Different Glucose Response Curves</span>: <span>A Latent Class Trajectory Analysis From</span> the <span>Prospective EGIR-RISC Study</span></a>. Diabetes Care. 2018 Aug;41(8):1740–8. </div>
</div>
<div id="ref-Ahlqvist2018" class="csl-entry" role="listitem">
<div class="csl-left-margin">9. </div><div class="csl-right-inline">Ahlqvist E, Storm P, Käräjämäki A, Martinell M, Dorkhan M, Carlsson A, et al. <a href="https://doi.org/10.1016/S2213-8587(18)30051-2">Novel subgroups of adult-onset diabetes and their association with outcomes: A data-driven cluster analysis of six variables</a>. The Lancet Diabetes &amp; Endocrinology. 2018 May;6(5):361–9. </div>
</div>
<div id="ref-Wagner2021" class="csl-entry" role="listitem">
<div class="csl-left-margin">10. </div><div class="csl-right-inline">Wagner R, Heni M, Tabák AG, Machann J, Schick F, Randrianarisoa E, et al. <a href="https://doi.org/10.1038/s41591-020-1116-9">Pathophysiology-based subphenotyping of individuals at elevated risk for type 2 diabetes.</a> Nature medicine. 2021 Jan;27(1):49–57. </div>
</div>
<div id="ref-Nicolaisen2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">11. </div><div class="csl-right-inline">Nicolaisen SK, Pedersen L, Witte DR, Sørensen HT, Thomsen RW. <a href="https://doi.org/10.1016/j.diabres.2023.110829"><span class="nocase">HbA1c-defined</span> prediabetes and progression to type 2 diabetes in <span>Denmark</span>: <span>A</span> population-based study based on routine clinical care laboratory data.</a> Diabetes research and clinical practice. 2023 Sep;203:110829. </div>
</div>
<div id="ref-Weller2020" class="csl-entry" role="listitem">
<div class="csl-left-margin">12. </div><div class="csl-right-inline">Weller BE, Bowen NK, Faubert SJ. <a href="https://doi.org/10.1177/0095798420930932">Latent <span>Class Analysis</span>: <span>A Guide</span> to <span>Best Practice</span></a>. Journal of Black Psychology. 2020 May;46(4):287–311. </div>
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/dp-next\.github\.io\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>